src/relay/transforms/concretize_like.cc

```diff
@@ -0,0 +1,160 @@
+    };
+    concrete_map_[Op::Get("collapse_sum_like")] = [](Expr data, Array<Integer> shape,
+                                                     DataType dtype) {
+      ICHECK_LE(shape.size(), std::numeric_limits<int64_t>::max());
+      static const Op& op = Op::Get("collapse_sum_to");
+      auto attrs = make_object<InitOpAttrs>();
+      auto cshape =
+          MakeConstantTensor(DataType::Int(32), {static_cast<int64_t>(shape.size())}, shape);
+      attrs->shape = shape;
+      return Call(op, {data, cshape}, Attrs(attrs));
+    };
+    concrete_map_[Op::Get("broadcast_to_like")] = [](Expr data, Array<Integer> shape,
+                                                     DataType dtype) {
+      return MakeBroadCastTo(data, shape);
+    };
+
+    for (const auto& pr : concrete_map_) {
+      if (!op_pat_.defined()) {
+        op_pat_ = IsExpr(pr.first);
+      } else {
+        op_pat_ = op_pat_ || IsExpr(pr.first);
+      }
+    }
+
+    data_pat_ = IsWildcard();
+    like_pat_ = IsWildcard();
+    unary_like_pat_ = (IsOp("zeros_like") || IsOp("ones_like"))({like_pat_});
+    binary_like_pat_ = (IsOp("reshape_like") || IsOp("collapse_sum_like") ||
+                        IsOp("broadcast_to_like"))({data_pat_, like_pat_});
+  }
+
+  Expr Callback(const Expr& pre, const Expr& post,
+                const Map<DFPattern, Array<Expr>>& node_map) const {
+    // we will rewrite iff the like argument has fully concrete shape
+    const CallNode* call_node = post.as<CallNode>();
+    ICHECK(call_node);
+    const OpNode* op_node = call_node->op.as<OpNode>();
+    ICHECK(op_node);
+    const Op op_ref = GetRef<Op>(op_node);
+    ICHECK(concrete_map_.count(op_ref) > 0);
+
+    Expr like = node_map[like_pat_][0];
+
+    if (!like->checked_type_.defined()) {
+      // TODO(@altanh): maybe because of the input being rewritten?
+      return post;
+    }
+
+    // skip trying to support this for now (ironic, as I was the one who added the feature)
+    if (const auto* attrs = call_node->attrs.as<ReshapeLikeAttrs>()) {
+      if (attrs->lhs_begin != 0 || attrs->rhs_begin != 0 || attrs->lhs_end.defined() ||
+          attrs->rhs_end.defined()) {
+        return post;
+      }
+    }
+
+    CHECK(like->checked_type_.defined())
+        << "ConcretizeLike requires checked types to be populated, please run type inference";
+    const TensorTypeNode* like_ty = like->checked_type().as<TensorTypeNode>();
+    ICHECK(like_ty) << "got non-Tensor argument type " << PrettyPrint(like->checked_type());
+
+    Array<Integer> cshape;
+    for (const auto& dim : like_ty->shape) {
+      if (const auto* imm = dim.as<IntImmNode>()) {
+        cshape.push_back(Integer(GetRef<IntImm>(imm)));
+        continue;
+      }
+      return post;
+    }
+
+    if (call_node->args.size() == 2) {
+      return concrete_map_.at(op_ref)(node_map[data_pat_][0], cshape, like_ty->dtype);
+    }
+    return concrete_map_.at(op_ref)(Expr(), cshape, like_ty->dtype);
```

Comment:
why empty Expr()?

Reply:
Yeah maybe this is too much of a hack, I'm just using it as a placeholder since the unary matches won't have a corresponding data Expr node. I'll rework this tmrw